src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1770,3 +1787,172 @@ HasUnresolvedExternParamsWalker(Node *expression, ParamListInfo boundParams)
 									  boundParams);
 	}
 }
+
+
+/*
+ * IsLocalReferenceTableJoin returns if the given query is a join between
+ * reference tables and local tables.
+ */
+static bool
+IsLocalReferenceTableJoin(Query *parse, List *rangeTableList)
+{
+	bool hasReferenceTable = false;
+	bool hasLocalTable = false;
+	ListCell *rangeTableCell = false;
+
+	bool hasReferenceTableReplica = false;
+
+	/*
+	 * We only allow join between reference tables and local tables in the
+	 * coordinator.
+	 */
+	if (!IsCoordinator())
+	{
+		return false;
+	}
+
+	/*
+	 * All groups that have pg_dist_node entries, also have reference
+	 * table replicas.
+	 */
+	PrimaryNodeForGroup(COORDINATOR_GROUP_ID, &hasReferenceTableReplica);
+
+	/*
+	 * If reference table doesn't have replicas on the coordinator, we don't
+	 * allow joins with local tables.
+	 */
+	if (!hasReferenceTableReplica)
+	{
+		return false;
+	}
+
+	if (FindNodeCheck((Node *) parse, QueryIsNotSimpleSelect))
+	{
+		return false;
+	}
+
+	foreach(rangeTableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+		DistTableCacheEntry *cacheEntry = NULL;
+
+		if (rangeTableEntry->rtekind == RTE_FUNCTION)
+		{
+			return false;
+		}
+
+		if (rangeTableEntry->rtekind != RTE_RELATION)
+		{
+			continue;
+		}
+
+
+		if (!IsDistributedTable(rangeTableEntry->relid))
+		{
+			hasLocalTable = true;
+			continue;
+		}
+
+		cacheEntry = DistributedTableCacheEntry(rangeTableEntry->relid);
+		if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
+		{
+			hasReferenceTable = true;
+		}
+		else
+		{
+			return false;
+		}
+	}
+
+	return hasLocalTable && hasReferenceTable;
+}
+
+
+/*
+ * QueryIsNotSimpleSelect returns true if node is a query which modifies or
+ * marks for modifications.
+ */
+static bool
+QueryIsNotSimpleSelect(Node *node)
```

Comment:
what about modifying CTEs?

Reply:
Modifying CTEs contain queries which are not-select. In `IsLocalReferenceTableJoin` we recursively search for those by using `FindNodeCheck`.